src/ipc/mod.rs

```diff
@@ -1,59 +1,328 @@
-    Grace(#[from] GraceError),
-}
-
-impl FriendlyError for ConnectionError {}
-
-#[derive(Debug, Clone, Error, Serialize, Deserialize)]
-#[error("Client performs a graceful shutdown.")]
-pub struct GraceError;
+#[derive(Debug, Clone)]
+pub struct ConnectionError;
 
-impl FriendlyError for GraceError {}
-
-pub fn listen_unix() -> impl Iterator<Item = RpcHandler> {
+pub fn listen_unix() -> impl Iterator<Item = ServerRpcHandler> {
     std::iter::from_fn(move || {
         let socket = self::transport::ServerSocket::Unix(self::transport::unix::accept());
-        Some(self::server::RpcHandler::new(socket))
+        Some(self::ServerRpcHandler::new(socket))
     })
 }
 
-pub fn listen_mmap() -> impl Iterator<Item = RpcHandler> {
+pub fn listen_mmap() -> impl Iterator<Item = ServerRpcHandler> {
     std::iter::from_fn(move || {
         let socket = self::transport::ServerSocket::Mmap(self::transport::mmap::accept());
-        Some(self::server::RpcHandler::new(socket))
+        Some(self::ServerRpcHandler::new(socket))
     })
 }
 
-pub fn connect_unix() -> self::transport::ClientSocket {
+pub fn connect_unix() -> ClientSocket {
     self::transport::ClientSocket::Unix(self::transport::unix::connect())
 }
 
-pub fn connect_mmap() -> self::transport::ClientSocket {
+pub fn connect_mmap() -> ClientSocket {
     self::transport::ClientSocket::Mmap(self::transport::mmap::connect())
 }
 
 pub fn init() {
     self::transport::mmap::init();
     self::transport::unix::init();
 }
+
+impl Drop for ClientRpc {
+    fn drop(&mut self) {
+        let socket = self.socket.take();
+        if let Some(socket) = socket {
+            if !std::thread::panicking() {
+                let mut x = CLIENTS.borrow_mut();
+                x.push(socket);
+            }
+        }
+    }
+}
+
+pub struct ClientRpc {
+    pub socket: Option<ClientSocket>,
+}
+
+impl ClientRpc {
+    fn new(socket: ClientSocket) -> Self {
+        Self {
+            socket: Some(socket),
+        }
+    }
+    fn _ok<U: Bincode>(&mut self, packet: U) -> Result<(), ConnectionError> {
+        self.socket.as_mut().unwrap().ok(packet)
+    }
+    fn _recv<U: Bincode>(&mut self) -> Result<U, ConnectionError> {
+        self.socket.as_mut().unwrap().recv()
+    }
+}
+
+static CLIENTS: PgRefCell<Vec<ClientSocket>> = unsafe { PgRefCell::new(Vec::new()) };
+
+pub fn client() -> Option<ClientRpc> {
+    if !crate::bgworker::is_normal() {
+        return None;
+    }
+    let mut x = CLIENTS.borrow_mut();
+    if let Some(socket) = x.pop() {
+        return Some(ClientRpc::new(socket));
+    }
+    let socket = match TRANSPORT.get() {
+        Transport::unix => connect_unix(),
+        Transport::mmap => connect_mmap(),
+    };
+    Some(ClientRpc::new(socket))
+}
+
+pub struct ServerRpcHandler {
+    socket: ServerSocket,
+}
+
+impl ServerRpcHandler {
+    pub(super) fn new(socket: ServerSocket) -> Self {
+        Self { socket }
+    }
+}
+
+macro_rules! define_packets {
```

Comment:
Long macro is really hard to understand. Do we have another way?

Reply:
#349